src/sqlancer/ASTBasedReducer.java

```diff
@@ -0,0 +1,470 @@
+package sqlancer;
+
+import net.sf.jsqlparser.expression.*;
+import net.sf.jsqlparser.expression.operators.relational.*;
+import net.sf.jsqlparser.parser.CCJSqlParserUtil;
+import net.sf.jsqlparser.statement.Statement;
+import net.sf.jsqlparser.statement.StatementVisitorAdapter;
+import net.sf.jsqlparser.statement.select.*;
+import sqlancer.common.query.Query;
+import sqlancer.common.query.SQLQueryAdapter;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.*;
+import java.util.function.BiConsumer;
+
+class ExpressionTransformer {
+
+    private static List<Expression> transformExpr(BinaryExpression expr) {
+        List<Expression> candidates = new ArrayList<>();
+        Expression lhs = (expr).getLeftExpression();
+        Expression rhs = (expr).getRightExpression();
+        candidates.add(lhs);
+        candidates.add(rhs);
+        return candidates;
+    }
+
+    private static List<Expression> transformExpr(Between expr) {
+        Expression lhs = expr.getBetweenExpressionStart();
+        Expression rhs = expr.getBetweenExpressionEnd();
+        return List.of(lhs, rhs);
+    }
+
+    public static List<Expression> candidateExpressions(Expression expr) {
+
+        // Queue<Expression> candidateExprQue = new PriorityQueue<>(Comparator.comparingInt(o ->
+        // o.toString().length()));
+        if (expr instanceof Parenthesis) {
+            // try removing a pair of brackets.
+            Parenthesis paren = (Parenthesis) expr;
+            return List.of(paren.getExpression());
+        } else if (expr instanceof BinaryExpression) {
+            return transformExpr((BinaryExpression) expr);
+        } else if (expr instanceof Between) {
+            return transformExpr((Between) expr);
+        } else if (expr instanceof LongValue) {
+            LongValue longValue = (LongValue) expr;
+            if (longValue.getValue() >= 1000) {
```

Comment:
Why are we checking for 1000? Would it not also make sense to reduce a value like 999?

Reply:
In fact, I plan to shorten the value and let them be less then 4 digits. So the constant values (including integer and float values) which are shorter than that would not be checked and reduced. 
Also, There is another aspect that needs to be considered:
``` java
do {
    observeChange = false;
    List<Expression> candidates = ExpressionTransformer.candidateExpressions(subExpr);
    for (Expression candidate : candidates) {
        try {
            setter.accept(parent, candidate);
            if (bugStillTriggers()) {
                subExpr = candidate;
                observeChange = true;
            }
        } catch (Exception ignoredException) {
        }
    }
    setter.accept(parent, subExpr);
} while (observeChange);
```

For an expression, An array of candidate expressions would be generated. (e.g. for a long value, it would be [null, 0, 1]). 

In the first round of while-loop, the expression would become `1`, but in the next round of loop, null, 0, 1 would be tried again respectively. The result would also be 1 and the `observeChange` would be true. Thus we've gone into an infinite loop. 

Therefore, we have to ensure that the constant values in the candidate list would not be reduced again. (negative long value should also be checkd, which I didn't do in this version.) 

Shortening those values longer than a certain length could be a simple solution to avoid infinite loop. Maybe we could discuss and come up with some other approaches for this problem.